### PR TITLE
crd for tiflash

### PR DESCRIPTION
**Pull request by DanielZhangQD**
### What problem does this PR solve? 
CRD for TiFlash
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
   manually deploy tidbcluster with CR

Code changes

 - Has Go code change

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support TiFlash in TidbCluster CR
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -278,6 +282,60 @@ type TiKVSpec struct {
...
+
+	// MaxFailoverCount limit the max replicas could be added in failover, 0 means no failover
+	// Optional: Defaults to 3
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	MaxFailoverCount *int32 `json:"maxFailoverCount,omitempty"`
+
+	// The DiskRequests of the persistent volume for TiFlash data storage.
+	// TiFlash supports multiple disks
+	DiskRequests []DiskRequests `json:"diskRequests"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/types.go:317]: I tend to use `StorageRequests` here

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 2]
```diff
@@ -278,6 +282,60 @@ type TiKVSpec struct {
...
+// +k8s:openapi-gen=true
+// TiDBSpec contains details of TiDB members
+type DiskRequests struct {
+	// The storageClassName of the persistent volume for data storage.
+	// Defaults to Kubernetes default storage class.
+	// +optional
+	StorageClassName *string `json:"storageClassName,omitempty"`
+
+	// StorageSize is the request storage size for TiFlash
+	StorageSize string `json:"storageSize"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/types.go:336]: We should keep the expression as same as `pd` and `tikv`, so i tend to use `storage` here.

File: docs/api-references/config.json [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+    "ProxyConfig",
+    "ClusterManagerPath",
+    "FlashStatus",
+    "FlashQuota",
+    "FlashUser",
+    "FlashProfile",
+    "FlashApplication",
+    "FlashProxy",
+    "FlashRaft",
+    "ClusterLog"
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply docs/api-references/config.json:25]: These are not all of the definitions of the types of TiFlash, just some of them, and I have tried `hideTypePatterns`, but they will still show in the doc.

File: docs/api-references/config.json [hunk 4]
```diff
@@ -26,4 +48,4 @@
     "k8s.io/apimachinery/pkg/apis/": "Kubernetes "
   },
   "markdownDisabled": false
-}
+}
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply docs/api-references/config.json:51]: updated

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 5]
```diff
@@ -278,6 +282,60 @@ type TiKVSpec struct {
...
+	// The DiskRequests of the persistent volume for TiFlash data storage.
+	// TiFlash supports multiple disks
+	DiskRequests []DiskRequests `json:"diskRequests"`
+
+	// Config is the Configuration of TiFlash
+	// +optional
+	Config *TiFlashConfig `json:"config,omitempty"`
+
+	// Dedicated TiDB cluster spec for TiFlash
+	TiDB *TiDBSpec `json:"tidb,omitempty"`
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:324]: Yes, should not mix it here, I will remove it.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:317]: updated

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go:336]: The `storage` for pd and tikv is configured in `corev1.ResourceRequirements` where the other fields are like `cpu`, `memory`, etc. However, `storage` here will not be clear enough.

**Comment by DanielZhangQD** [comment]
@Yisaer comments addressed, PTAL again.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 2 above)
Comment (Yisaer) [reply pkg/apis/pingcap/v1alpha1/types.go:333]: Event if assuming the `storage` is not clear enough, I think there is no need to make difference between `pd`/`tikv`/`tiflash` and kubernetes resources expression. `StorageSize` seems like the property to dedicate the byte size of the storage requests which is common used in `TiDB`.

File: pkg/apis/pingcap/v1alpha1/tiflash_config.go [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+package v1alpha1
+
+// Port from TiFlash configurations till 2020/04/02
+
+// TiFlashConfig is the configuration of TiFlash.
+// +k8s:openapi-gen=true
+type TiFlashConfig struct {
+	// commonConfig is the Configuration of TiFlash process
+	// +optional
+	CommonConfig *CommonConfig `json:"config,omitempty"`
```
Comment (weekface) [inline pkg/apis/pingcap/v1alpha1/tiflash_config.go:23]: Do we need add `toml` to the tag?

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tiflash_config.go:23]: Good catch, I will update them.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 2 above)
Comment (cofyc) [reply pkg/apis/pingcap/v1alpha1/types.go:333]: how about using a more generic and Kubernetes-style to configure volumes starting from tiflash? we can use volumeClaimTemplate to configure volumes for container and volumeMount to configure how to mount them. then we can use all functions they provide, e.g.

- `volumeMode`, `selector` from `volumeClaimTemplate`.
- `subPath` from `volumeMount` (https://github.com/pingcap/tidb-operator/issues/1525#issuecomment-602016020)

Comment (Yisaer) [reply pkg/apis/pingcap/v1alpha1/types.go:333]: Good idea!

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/types.go]: Updated.

**Comment by DanielZhangQD** [comment]
@weekface @cofyc @Yisaer Code updated as we have discussed, PTAL again.

**Pull request merged**